crates/bcr-wdc-treasury-service/src/debit/service.rs

```diff
@@ -0,0 +1,303 @@
+}
+
+#[derive(Clone)]
+pub struct Service<Wlt, ProofCl> {
+    pub wallet: Wlt,
+
+    pub proof: ProofCl,
+
+    pub secp_ctx: bitcoin::secp256k1::Secp256k1<bitcoin::secp256k1::SignOnly>,
+    pub signing_keys: bitcoin::secp256k1::Keypair,
+}
+
+impl<Wlt, ProofCl> Service<Wlt, ProofCl>
+where
+    Wlt: Wallet,
+{
+    pub async fn mint_from_ebill(
+        &self,
+        ebill_id: String,
+        amount: Amount,
+    ) -> Result<cdk::wallet::MintQuote> {
+        let request = web::signatures::RequestToMintFromEBillDesc { ebill: ebill_id };
+        let borshed = borsh::to_vec(&request).map_err(Error::BorshIO)?;
+        let msg = bcr_wdc_keys::into_secp256k1_msg(&borshed);
+        let signature = self.secp_ctx.sign_schnorr(&msg, &self.signing_keys);
+        let signed_request = web::signatures::SignedRequestToMintFromEBillDesc {
+            data: request,
+            signature,
+        };
+        self.wallet.mint_quote(amount, signed_request).await
+    }
+}
+
+impl<Wlt, ProofCl> Service<Wlt, ProofCl>
+where
+    Wlt: Wallet,
+    ProofCl: ProofClient,
+{
+    pub async fn redeem(
+        &self,
+        inputs: Vec<cdk00::Proof>,
+        outputs: Vec<cdk00::BlindedMessage>,
+    ) -> Result<Vec<cdk00::BlindSignature>> {
+        // cheap verifications
+        // 1. no empty inputs or outputs
+        if inputs.is_empty() || outputs.is_empty() {
+            return Err(Error::EmptyInputsOrOutputs);
+        }
+        // 2. no zero amounts in inputs or outputs
+        let zero_outputs = outputs.iter().any(|output| output.amount == Amount::ZERO);
+        let zero_inputs = inputs.iter().any(|output| output.amount == Amount::ZERO);
+        if zero_outputs || zero_inputs {
+            return Err(Error::ZeroAmount);
+        }
+        // 3. inputs and outputs have equal amounts
+        let total_input = inputs
+            .iter()
+            .fold(Amount::ZERO, |total, proof| total + proof.amount);
+        let total_output = outputs
+            .iter()
+            .fold(Amount::ZERO, |total, proof| total + proof.amount);
+        if total_input != total_output {
+            return Err(Error::UnmatchingAmount(total_input, total_output));
+        }
+        // expensive verifications
+        // 1. outputs' keyset ID(s) are active
+        let unique_ids: Vec<_> = outputs.iter().map(|p| p.keyset_id).unique().collect();
+        let infos = self.wallet.get_keysets_info(&unique_ids).await?;
+        for info in infos {
+            if !info.active {
+                return Err(Error::InactiveKeyset(info.id));
+            }
+        }
+        // 2. burning crsat, implicitly checking proofs
+        self.proof.burn(inputs.clone()).await?;
+
+        let signatures = match self.wallet.swap_to_messages(outputs).await {
+            Ok(signatures) => signatures,
+            Err(e) => {
+                // restore the burned proofs
+                self.proof.restore(inputs).await?;
```

Comment:
What happens if the restoration fails? :thinking: Is it relevant?
Maybe a `warn!` log or something would be useful?